geometry/parse-ply.cpp

```diff
@@ -0,0 +1,316 @@
+    {
+        LOGTHROW(err4, std::runtime_error)
+            << "Error parsing element defined by: " << line;
+    }
+
+    elements.emplace_back(name, count);
+    PlyElement& el = elements.back();
+    // parse the element properties
+    do
+    {
+        nextLine(f, line);
+    }
+    while (parseProperty(line, el.props) || parseComment(line, comments));
+
+    if (el.props.empty())
+    {
+        LOGTHROW(err4, std::runtime_error)
+            << "Encountered element with no properties: " << name;
+    }
+
+    return true;
+}
+
+/// Check "format" line in PLY file, throws on unexpected inputs
+void checkFormatLine(std::ifstream& f)
+{
+    std::string line;
+    nextLine(f, line); // read whole line
+
+    std::string keyword, format, version;
+    if (!(std::istringstream(line) >> keyword >> format >> version))
+    {
+        LOGTHROW(err4, std::runtime_error)
+            << "Error parsing format line: " << line;
+    }
+    if (keyword != "format")
+    {
+        LOGTHROW(err4, std::runtime_error)
+            << "Expected the second line to contain \"format\" specification";
+    }
+    if (format != "ascii")
+    {
+        LOGTHROW(err4, std::runtime_error)
+            << "Only ASCII format is supported, got: " << format;
+    }
+}
+
+/// Parses the header section of PLY file and returns a vector of elements with
+/// their properties and comments
+std::pair<std::vector<PlyElement>, std::vector<std::string>>
+    parseHeader(std::ifstream& f)
+{
+    // check magic number
+    std::string line; 
+    nextLine(f, line);
+    if (line != "ply")
+    {
+        LOGTHROW(err4, std::runtime_error)
+            << "Unexpected magic number of PLY file: " << line;
+    }
+
+    // check line specifying PLY format ("format ascii 1.0")
+    checkFormatLine(f); 
+    
+    // parse elements & comments
+    std::vector<PlyElement> elements;
+    std::vector<std::string> comments;
+    nextLine(f, line);
+    do
+    {
+        if (parseElement(f, line, elements, comments))
+        {
+            continue; // parseElement() jumps to the line after element
+                      // definition
+        }
+        else if (!parseComment(line, comments))
+        {
+            LOGTHROW(err4, std::runtime_error)
+                << "Encountered unexpected line in header (not a comment nor "
+                   "element definition): "
+                << line;
+        }
+
+        nextLine(f, line);
+    }
+    while (!ba::starts_with(line, "end_header"));
+
+    return std::make_pair(std::move(elements), std::move(comments));
+}
+
+
+void parsePly(PlyParserBase& parser, const fs::path& path)
+{
+    LOG(info1) << "Loading PLY file from " << path << ".";
+    std::ifstream f(path.native());
+    if (!f.good()) {
+        LOGTHROW(err4, std::runtime_error)
+            << "Unable to open file " << path << ".";
+    }
+    // not setting failbit - is set on eof
+    f.exceptions(std::ios::badbit);
```

Comment:
`std::getline` sets failbit along with eof when reaching the end of file. I will surround it with try-catch as suggested in: https://stackoverflow.com/questions/50662874/getline-setting-failbit-along-with-eof

Reply:
Are you using `std::getline` anywhere else except parsing of the header (where the EOF shouldn't happen anyway)?

Reply:
Not anymore (after https://github.com/melowntech/libgeometry/pull/14/commits/f082c16d6e9dc6cab1e1a3ca0ce2952f7f0eb3b1 where I started passing `istream` instead of `string` to `PlyParserBase`) so it was easy to change in `nextLine()`.